File: 06-greengrassguard-predictive-maintenance/sensors/mqtt_publisher.py
"""Publishes sensor readings to the exact MQTT topic pattern the fog layer subscribes to."""
import json

TOPIC_TEMPLATE = "greengrassguard/{assetId}/{metric}"


def publish_reading(mqtt_client, reading):
    """Publish a reading dict as JSON to greengrassguard/{assetId}/{metric}.

    reading must contain at least assetId and metric; the whole dict is the payload
    so vibration readings' extra `window` field passes through untouched.
    """
    topic = TOPIC_TEMPLATE.format(assetId=reading["assetId"], metric=reading["metric"])
    payload = json.dumps(reading)
    mqtt_client.publish(topic, payload)
    return topic

File: 08-officeiq-smart-office-comfort/sensors/mqttPublisher.js
'use strict';

// topic pattern is the fog layer's only subscription filter, so it must match the contract exactly
function topicFor(reading) {
  return `officeiq/${reading.zoneId}/${reading.metric}`;
}

function publishReading(mqttClient, reading) {
  const topic = topicFor(reading);
  const payload = JSON.stringify(reading);
  mqttClient.publish(topic, payload);
}

module.exports = { publishReading, topicFor };
